contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,1096 @@
+         || locator.withdrawalVault() != address(_withdrawalVault)
+        ) {
+            revert IncorrectLocatorAddresses();
+        }
+    }
+
+    function _assertFeeDistribution() internal view {
+        (uint16 treasuryFeeBasisPoints, , uint16 operatorsFeeBasisPoints) = _lido.getFeeDistribution();
+        // New fee values for staking module are set as values relative to the all rewards gained
+        // Although Lido.getFeeDistribution() returns values relative to total fee taken by the protocol from all rewards
+        // So need to convert "relative" hardcoded values into their "absolute" form
+        uint256 totalFee = _lido.getFee();
+        if (
+            (NOR_STAKING_MODULE_MODULE_FEE_BP * TOTAL_BASIS_POINTS) / totalFee != operatorsFeeBasisPoints
+         || (NOR_STAKING_MODULE_TREASURY_FEE_BP * TOTAL_BASIS_POINTS) / totalFee != treasuryFeeBasisPoints
+        ) {
+            revert IncorrectFeeDistribution();
+        }
+    }
+
+    function _assertCorrectDSMParameters() internal view {
+        IDepositSecurityModule dsm = _depositSecurityModule;
+        if (
+            dsm.getMaxDeposits() != EXPECTED_DSM_MAX_DEPOSITS_PER_BLOCK
+         || dsm.getPauseIntentValidityPeriodBlocks() != EXPECTED_DSM_PAUSE_INTENT_VALIDITY_PERIOD_BLOCKS
+         || dsm.getMinDepositBlockDistance() != EXPECTED_DSM_MIN_DEPOSIT_BLOCK_DISTANCE
+        ) {
+            revert IncorrectDepositSecurityModuleParameters(address(dsm));
+        }
+    }
+
+    function _upgradeProxyImplementations() internal {
+        _upgradeOssifiableProxy(_locator, _locatorImplementation);
+        _upgradeOssifiableProxy(_accountingOracle, _accountingOracleImplementation);
+        _upgradeOssifiableProxy(_validatorsExitBusOracle, _validatorsExitBusOracleImplementation);
+        _upgradeOssifiableProxy(_stakingRouter, _stakingRouterImplementation);
+        _upgradeOssifiableProxy(_withdrawalQueue, _withdrawalQueueImplementation);
+    }
+
+    function _upgradeOssifiableProxy(IOssifiableProxy proxy, address newImplementation) internal {
+        // NB: Such separation of external call into a separate function saves contract bytecode size
+        proxy.proxy__upgradeTo(newImplementation);
+    }
+
+    function _assertAdminsOfProxies(address admin) internal view {
+        _assertProxyAdmin(_locator, admin);
+        _assertProxyAdmin(_accountingOracle, admin);
+        _assertProxyAdmin(_stakingRouter, admin);
+        _assertProxyAdmin(_validatorsExitBusOracle, admin);
+        _assertProxyAdmin(_withdrawalQueue, admin);
+    }
+
+    function _assertProxyAdmin(IOssifiableProxy proxy, address admin) internal view {
+        if (proxy.proxy__getAdmin() != admin) revert IncorrectProxyAdmin(address(proxy));
+    }
+
+    function _assertOracleReportSanityCheckerRoles() internal view {
+        IOracleReportSanityChecker checker = _oracleReportSanityChecker;
+        // NB: oracleReportSanityChecker admin is set to agent upon deployment at once
```

Comment:
It is actually set upon deployment, because admin is passed as an arg of OracleReportSanityChecker ctor. Anyway removed the comments to avoid ambiguity.